### ISSUE TITLE
[Due for payment 2025-04-02] [Performance][Search] Opening search for the first time on a big account is causing the app to hang massively

### ISSUE THREAD

**Issue by hannojg** [Expensify/App#46590]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

## What performance issue do we need to solve?

In a customers meeting the customer was trying to open the search for the first time in NewDot. It caused the app to hang up, and it was a very bad user experience (see the recording):

https://github.com/user-attachments/assets/99d97430-9545-475a-afca-7d0c5b5d1437

## What is the impact of this on end-users?

User perception of the app is very bad, as it's super laggy, basically not usable (the browser even warns that the app is hanging).

## List any benchmarks that show the severity of the issue

The customer shared a profile trace with us:

[Firefox 2024-07-25 10.42 profile.json.gz](https://github.com/user-attachments/files/16443244/Firefox.2024-07-25.10.42.profile.json.gz)

(note: the trace also contains other test cases as well)

The customer had ~15k reports loaded in onyx when these tests were conducted, although in focus mode only 6 chats were displayed.

## Proposed solution (if any)

None yet, I will go through the profile and see what can be optimised, what exactly caused those lags.

## List any benchmarks after implementing the changes to show impacts of the proposed solution (if any)

not available yet 

## Platforms:

Which of our officially supported platforms is this issue occurring on?
- [ ] Android: Native
- [ ] Android: mWeb Chrome
- [ ] iOS: Native
- [ ] iOS: mWeb Safari
- [x] MacOS: Chrome / Safari
- [ ] MacOS: Desktop

**Version Number:** v9.0.11-5
**Reproducible in staging?:** not tested
**Reproducible in production?:** yes
**Email or phone of affected tester (no customers):** customer
**Logs:** See performance file
**Notes/Photos/Videos:** See attached video
**Expensify/Expensify Issue URL:** n/a
**Issue reported by:** @hannojg
**Slack conversation:** https://expensify.slack.com/archives/C05LX9D6E07/p1721919928992729

[View all open jobs on Upwork](https://www.upwork.com/ab/jobs/search/?q=Expensify%20React%20Native&sort=recency&user_location_match=2)

**Comment by hannojg** [comment]
cc @sakluger

**Comment by marcaaron** [comment]
Curious what is happening with this especially to see if it is related to large numbers of reports (seems likely as there are 15k reports). Gonna assign myself in case @hannojg needs any assistance.

**Comment by sakluger** [comment]
No updates

**Comment by sakluger** [comment]
No update

**Comment by hannojg** [comment]
A large part of this hanging is due to us building the search options the moment we open the search page:

- https://github.com/Expensify/App/blob/4bac1000e36d0a8ba5fba945e87053d7b8cf2ef2/src/pages/ChatFinderPage/index.tsx#L78-L91

I am thinking about how to solve this here, by moving the operations in a first step to a separate thread/web worker:

- https://expensify.slack.com/archives/C05LX9D6E07/p1723454316012939?thread_ts=1723209631.007999&cid=C05LX9D6E07

**Comment by hannojg** [comment]
I just opened a PR hat shows a PoC for using web workers for running the search, which will unblock the main thread:

- https://github.com/Expensify/App/pull/47432

I am now seeking approval on this new technical design [here](https://expensify.slack.com/archives/C05LX9D6E07/p1723649031113319?thread_ts=1723209631.007999&cid=C05LX9D6E07).

(Note: I am OOO for the rest of the week and will continue next week Monday 19.08)

**Comment by sakluger** [comment]
Moving to weekly for now since Hanno is out until Monday.

**Comment by chrispader** [comment]
Commenting for assignment :)

**Comment by sakluger** [comment]
Thanks for helping, @chrispader. I'll be OOO for the next week. Hopefully we'll have some more details from the investigation when I return!

**Comment by hannojg** [comment]
We were able to reproduce the bigger lags when running on a virtual machine, where the hardware is more constraint than on our recent MacBook Pro models directly. (note: we tested against a recent adhoc build that contained some search improvements).

The thing that takes the most time when opening the search page / chat finder page is:
- calling `createOptionsList`
- calling `getSearchOptions` (on the results of createOptionsList)
- calling `triggerUnreadUpdate` also took ~1-2 seconds 

![Screenshot 2024-08-22 at 16 56 57](https://github.com/user-attachments/assets/fcc9b6d1-1939-4545-9c89-89821e1e0718)

![Screenshot 2024-08-22 at 16 57 03](https://github.com/user-attachments/assets/6802ccab-fc4a-4f0f-89eb-defb1978ee7e)

![Screenshot 2024-08-22 at 16 58 09](https://github.com/user-attachments/assets/98cc3258-2e76-4078-b38e-96295a3e7209)

![Screenshot 2024-08-22 at 16 58 15](https://github.com/user-attachments/assets/5b4cad7d-5df3-40fb-9bbc-b1f05cc388e0)

We will come up with solutions for improving the performance of these functions (in the current single threaded architecture).
(Some of these solutions might blend with the solutions we are working on [here](https://github.com/Expensify/App/issues/46591#issuecomment-2273761808))

**Comment by sakluger** [comment]
@hannojg any updates on this investigation?

**Comment by sakluger** [comment]
No updates - seems like we are focusing on improving search speed first. There is a WIP PR for that (https://github.com/Expensify/App/pull/48652), hopefully we'll get back to this one after that.

**Comment by sakluger** [comment]
No updates.

**Comment by marcaaron** [comment]
What's the latest on this one?

**Comment by hannojg** [comment]
We were able to get another, cleaner, sample of that problem here which relates to creating the options (which causes the lag):

- https://github.com/Expensify/App/issues/46593#issuecomment-2387772633

I am sure now that we have a cleaner recording we can come up with good solutions to make this faster!

**Comment by sakluger** [comment]
No updates.

**Comment by sakluger** [comment]
No updates.

**Comment by sakluger** [comment]
No updates

**Comment by sakluger** [comment]
No updates

**Comment by sakluger** [comment]
No updates

**Comment by sakluger** [comment]
No updates

**Comment by sakluger** [comment]
No update

**Comment by chrispader** [comment]
I will try to look into this issue this week!

**Comment by hannojg** [comment]
We've identified one place where the performance can be improved.
When opening the search page we need to construct out search options. Part of that step is to order the options.

Opening the full search page takes on my Mac:

```
Timing:development.new.expensify.open_search 416.6999999284744
```

For that, we need to create the search options, which makes up for ~70ms of the 416ms:
```
Timing:development.new.expensify.load_search_options 37.700000047683716
Timing:development.new.expensify.load_search_options 34.200000047683716
```

Of those 70ms half of the time is spend on ordering the items.

We use lodash order functions, [which are known to be slow](https://www.measurethat.net/Benchmarks/ShowResult/577003). We should replace the lodash sort with our own function that is optimized for performance.

---

Note: the numbers may not seem that drastically as those are from my high end MacBook. The numbers for android on a lower end device are:

```
Timing:development.new.expensify.load_search_options 790.5101539939642
Ordering options took 475.9423850029707ms
Timing:development.new.expensify.open_search 6194.899076998234
```

---

This is only one part of where the time goes for creating the search options, we will continue to where the other time is spend and how it can be optimized.

**Comment by hannojg** [comment]
@sakluger can we please reopen? the PR only addressed the issue partially

**Comment by hannojg** [comment]
One of the main PRs was merged. The game plan to finish this is now:

- Get this one merged: https://github.com/Expensify/App/pull/55298
- Further optimize smaller code pieces for performance in OptionListUtils
- We call `getValidOptions` many times and in performance profiles we see that this calls `shouldReportBeInOptionList` which is quite expensive. We think this can be optimized. We should calculate a bitwise flag for the reports once (in the search context). Then we can quickly and as often as we like check if a report should be displayable
- Right now there is still an inefficient where the FastSearch gets recreated every time the reports change. This can be optimized to only update the FastSearch/SearchTree, which will be significantly faster than creating it from scratch. Issue ticket for that is here:
	- https://github.com/Expensify/App/issues/51150

after that our search should be 🚀🚀🚀

**Comment by sakluger** [comment]
Hey @hannojg - fyi, if you don't want an issue to auto-close, just remove the `#` before the issue URL on the linked PR. We have automation set up to auto-close GH issues when you link it in the PR using this format:

<img width="886" alt="Image" src="https://github.com/user-attachments/assets/a7bded60-6f08-45c6-85d4-621f8614a177" />

**Comment by mallenexpensify** [comment]
This is on #quality, unsure of next steps, looks like it's moving along, I made it status of `LOW` for now so that I don't have it to categorize each week.

**Comment by hannojg** [comment]
Yes, we are working on different things to improve the performance of the search.
A good roadmap plan can be read here:

- https://github.com/Expensify/App/issues/46590#issuecomment-2595568361

**Comment by sakluger** [comment]
Commenting for Melvin since I had the assignment star.

**Comment by luacmartins** [comment]
We just merged a PR

**Comment by sakluger** [comment]
Awesome to see this one completed! No payment required through Upwork or NewDot payments.

**Comment by garrettmknight** [comment]
$250 approved for @Ollyws
